rslib/src/sched.rs

```diff
@@ -42,15 +42,34 @@ fn rollover_for_today(
 
 /// The number of times the day rolled over between two timestamps.
 fn days_elapsed(start: i64, end: i64, rollover_today: i64) -> u32 {
-    // get the number of full days that have elapsed
-    let secs = (end - start).max(0);
-    let days = (secs / 86_400) as u32;
+    let start_dt = Local.timestamp(start, 0);
```

Comment:
Ah, my second point was incorrect - as you're using a timezone rather than a fixed offset, the starting cutoff should not vary throughout the year.

The issue remains that the code needs to have all inputs passed into it. AnkiWeb could potentially be storing a timezone instead of a fixed offset which would allow an approach like this to work, but I'd like to better understand the problem you're trying to solve before we go down that path.

Reply:
The core issue is: what problem am I trying to fix. So, in response to this and your first comment above, I'll try to explain here, then submit specific tests without changes to the implementation to clarify.

If I am running Anki around rollover time, I see all my day counts (new, learned and review) are zeroed at rollover time as expected but new reviews don't appear until an hour later. My testing lead me to look at the calculation of days elapsed and my change to the python implementation (previously submitted PR) solved the problem for me by calculating days elapsed mush as in the PR.

There is a complicating factor that in my collection crt is 3am, 24 April 2019, rather than 4am. Looking a code in git, including history, I don't see how this can be. it looks like it should be 4am. And it looks like the relevant code has been unchanged since long before I first used Anki. I was running Anki from Linux Mint repo at the time, but I have since removed it, so I can't be sure what version it was.  The Mint repo currently has version 2.1.0+dfsg~b36-1 but I don't now how to determine if it has been updated in the meantime. And I can't reproduce initial conditions without setting up a test system with date set back to April 2019. Anyway, I might try installing what is currently available and see what happens when I set up a new collection. 

I don't think crt being 3am is the root cause of the delay I am seeing in reviews coming available after rollover time, but I may be mistaken. Perhaps counts are zeroed at 3am and new cards come available at 4am.

I will stop with changes to the implementation and work on tests to clarify the problem. It may take me a while, but I'll come back with an update...

Reply:
Thanks for elaborating. I managed to recreate the problem you describe using the following failing test:

```rust
        // collection created at 3am on the 6th, so day 1 starts at 4am on the 7th, and day 3 on the 9th.
        let crt = mdt.ymd(2018, 8, 6).and_hms(3, 0, 0).timestamp();

        let mst_offset = mst.utc_minus_local() / 60;
        let now = mst.ymd(2018, 8, 9).and_hms(1, 59, 59).timestamp();
        assert_eq!(elap(crt, now, mst_offset, 4), 2);
        let now = mst.ymd(2018, 8, 9).and_hms(3, 59, 59).timestamp();
        assert_eq!(elap(crt, now, mst_offset, 4), 2);
        let now = mst.ymd(2018, 8, 9).and_hms(4, 0, 0).timestamp();
        assert_eq!(elap(crt, now, mst_offset, 4), 3);
```

So in this case the extra checks were required :-) I'll look into this further.